#' plotcp function
#'
#' Draws plot of copy number matrix
#'
#' @param mae MultiAssayExperiment object. Must contain assays GISTIC or a matrix of copy number values, assay should contain "GISTIC" in the name,
#' and Cohort (from `rnasubset` or `mutsubset`)
#' @param genecp data frame generated by diffcp function
#' @param gene character(1) gene of interest
#' @param n numeric(1) number of genes to plot default is 20
#'
#' @import data.table
#'
#' @return interactive oncoprint-like plot for top 20 differentially copy number altered genes
#'
#' @examples
#' library(curatedTCGAData)
#' library(TCGAutils)
#' lusc <- curatedTCGAData("LUSC", c("Mutation", "RNASeq2GeneNorm", "GISTIC_ThresholdedByGene"), FALSE)
#' genome(lusc[[2]]) <- vapply(genome(lusc[[2]]), translateBuild, character(1L))
#' seqlevelsStyle(lusc[[2]]) <- "NCBI"
#' lusc2 <- qreduceTCGA(lusc)
#'
#' #remove normal samples and just focus on primaries
#' tnmae <- splitAssays(lusc2, c("01", "11"))
#'
#' lusc.t <- tnmae[, , grep("^01", names(tnmae))]
#' #split tumor and normal samples
#' lusc_tn <- splitAssays(lusc2, c("01", "11"))
#' #remake MultiAssayExperiment with only primary tumor samples
#' lusc_t <- lusc_tn[, , grep("^01", names(lusc_tn))]
#' lusc_egfr <- rnasubset(lusc_t, "EGFR", 10)
#' egfr_dcp <- diffcp(lusc_egfr)
#' plotcp(lusc_t.egfr, egfr_dcp, "EGFR", 20)
#'
#' @export
#'
plotcp <- function(mae, genecp, gene, n = 20) {
  stopifnot(class(mae) == "MultiAssayExperiment", any(grepl("GISTIC", names(mae))), any(grepl("Cohort", names(mae))))
  if(length(grep("GISTIC", names(mae))) > 1)
    stop("Only one copy number assay is allowed")
  if(length(grep("Cohort", names(mae))) > 1)
    stop("Only one cohort comparison is allowed")

  # Find which assays contain mutation and expression level
  cp_assay <- grep("GISTIC", names(mae))
  exp_assay <- grep("Cohort", names(mae))

  mae2 <- intersectColumns(mae[, , c(cp_assay, exp_assay)])

  annot <- dcast(as.data.frame(sampleMap(mae2)), primary ~ assay, value.var = "colname")
  lvl2 <- data.frame(Cohort = colnames(mae2[[2]]), Level = mae2[[2]][1,])
  lvl3 <- merge(annot, lvl2, by = "Cohort")

  lvl.high <- lvl3[lvl3$Level == "high", grep("GISTIC", colnames(lvl3))]
  lvl.low <- lvl3[lvl3$Level == "low", grep("GISTIC", colnames(lvl3))]

  cp <- assay(mae2[[1]])

  #Choose which genes to plot (num_genes)
  glist <- c(genecp$Gene[1:n], gene)
  glist <- glist[!duplicated(glist)]

  #limit mut matrix to genes chosen above
  cp.high <- cp[glist, lvl.high]
  cp.low <- cp[glist, lvl.low]
  #getting gene and patient order for plotting
  cp1.melt.high <- melt(cp.high)
  gg1.high <- dcast(cp1.melt.high, Var2 ~ Var1)
  setDT(gg1.high)
  gene.order.high <- names(sort(apply(gg1.high[,colnames(gg1.high)[-1], with=F], 2, sum), decreasing=T))
  setkeyv(gg1.high, gene.order.high)
  cp1.melt.high$level <- "high"

  cp1.melt.low <- melt(cp.low)
  gg1.low <- dcast(cp1.melt.low, Var2 ~ Var1)
  setDT(gg1.low)
  gene.order.low <- names(sort(apply(gg1.low[,colnames(gg1.low)[-1], with=F], 2, sum), decreasing=T))
  setkeyv(gg1.low, gene.order.low)
  cp1.melt.low$level <- "low"

  #combine mutation groups order patients according to sorting from individual dcasts
  cp1.melt <- rbind(cp1.melt.high, cp1.melt.low)
  setnames(cp1.melt, c("Gene", "name", "value", "level"))
  cp1.melt$name <- factor(cp1.melt$name,
                          levels = c(as.character(rev(gg1.high$Var2)),
                                     as.character(rev(gg1.low$Var2))))
  cp1.melt$Gene <- factor(cp1.melt$Gene, levels = rev(gene.order.high))

  cg1 <- ggplot(cp1.melt, aes(name, Gene, fill = as.factor(value))) +
    geom_tile(colour=c("white")) +
    labs(x = "Patient", y="Gene") +
    scale_x_discrete(expand=c(0,0))+scale_y_discrete(expand=c(0,0)) +
    scale_fill_manual(values = c("#0571b0", "#92c5de", "#f7f7f7", "#f4a582", "#ca0020"),
                      labels = c(-2:2),
                      guide = guide_legend(title = "GISTIC score")) +
    theme(title = element_text(size=16),
          axis.ticks = element_blank(),
          axis.text.x = element_blank(),
          axis.text.y = element_text(size=12),
          axis.title.x = element_text(size=14),
          axis.title.y = element_text(size=14),
          legend.text = element_text(size=12)) +
    ggtitle(paste(gene)) +
    facet_grid(. ~ level, scales = "free", space = "free")
  cg1
  #ggplotly(cg1)
}



